### PR TITLE
Model Adapter

### PR DESCRIPTION
**Pull request by akleeman**
This change adds the ability to make a `RegressionModel<FeatureTypeX>` that effectively wraps a `RegressionModel<FeatureTypeY>` by providing a function which converts from `FeatureTypeY -> FeatureTypeX`.

This is done using a new class `AdaptedRegressionModel` which is then used to build `LinearRegression` out of `LeastSquaresRegression`.

Additionally adds some further testing which allows very basic performance testing (right now on a simple linear dataset).

### CODE REVIEW

File: albatross/core/model_adapter.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+ * This provides a way of creating a RegressionModel<X> which
+ * wraps a RegressionModel<Y> as long as you provide a way of
+ * converting from X to Y.
+ *
+ * A good example can be found in the definition of LinearRegression.
+ */
+template <typename FeatureType, typename SubModelType>
+class AdaptedRegressionModel : public RegressionModel<FeatureType> {
+ public:
+  typedef typename SubModelType::Feature SubFeature;
```
Comment (benjamin0) [inline albatross/core/model_adapter.h]: nit: `using`?

File: albatross/core/model_adapter.h [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+  AdaptedRegressionModel() : sub_model_(){};
+  AdaptedRegressionModel(SubModelType& sub_model) : sub_model_(sub_model){};
```
Comment (benjamin0) [inline albatross/core/model_adapter.h]: const ref?

File: albatross/core/model_adapter.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+  ParameterStore get_params() const override { return sub_model_.get_params(); }
+
+  void unchecked_set_param(const std::string& name,
+                           const double value) override {
+    sub_model_.set_param(name, value);
+  }
+
+  template<class Archive>
+  void save(Archive & archive) const
```
Comment (benjamin0) [inline albatross/core/model_adapter.h]: Looks like this file may need to be clang-formatted.

File: albatross/models/least_squares.h [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef ALBATROSS_MODELS_LEAST_SQUARES_H
+#define ALBATROSS_MODELS_LEAST_SQUARES_H
+
+/*
+ * Here we define a LeastSquaresRegression model which is less because
```
Comment (benjamin0) [inline albatross/models/least_squares.h]: less...?

File: albatross/models/least_squares.h [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+  LeastSquaresRegression() {};
+  std::string get_name() const override { return "least_squares"; };
+
+ protected:
+
+  /*
+   * This lets you customize the least squares approach if need be,
+   * default uses the QR decomposition.
+   */
+  virtual Eigen::VectorXd least_squares_solver(Eigen::MatrixXd &A, Eigen::VectorXd &b) {
```
Comment (benjamin0) [inline albatross/models/least_squares.h]: Can these be const?

File: albatross/core/model_adapter.h (hunk 3 above)
Comment (akleeman) [reply albatross/core/model_adapter.h]: yeah, adding auto clang formatting is on my TODO list.  I'll run it on the full repo soon.

File: albatross/core/model_adapter.h (hunk 2 above)
Comment (akleeman) [reply albatross/core/model_adapter.h]: nice catch

File: albatross/core/model_adapter.h (hunk 1 above)
Comment (akleeman) [reply albatross/core/model_adapter.h]: done

File: albatross/models/least_squares.h (hunk 4 above)
Comment (akleeman) [reply albatross/models/least_squares.h]: that was just generally a confusing out of date comment, I removed it entirely.

**Pull request merged**